server/commands/connected_objects.go

```diff
@@ -0,0 +1,351 @@
+) error {
+
+	depth, ok := graph.ResolutionDepthFromContext(ctx)
+	if !ok {
+		ctx = graph.ContextWithResolutionDepth(ctx, c.ResolveNodeLimit)
+	} else {
+		if depth == 0 {
+			return serverErrors.AuthorizationModelResolutionTooComplex
+		}
+
+		ctx = graph.ContextWithResolutionDepth(ctx, depth-1)
+	}
+
+	storeID := req.StoreID
+
+	targetUserType, _ := tuple.SplitObject(req.User.Object)
+
+	targetUserRef := &openfgapb.RelationReference{
+		Type:     targetUserType,
+		Relation: req.User.Relation,
+	}
+
+	sourceObjRef := &openfgapb.RelationReference{
+		Type:     req.ObjectType,
+		Relation: req.Relation,
+	}
+
+	// build the graph of possible edges between object types in the graph based on the authz model's type info
+	g := graph.BuildConnectedObjectGraph(c.Typesystem)
+
+	// find the possible incoming edges (ingresses) between the target user reference and the source (object, relation) reference
+	ingresses, err := g.RelationshipIngresses(sourceObjRef, targetUserRef)
+	if err != nil {
+		return err
+	}
+
+	for _, ingress := range ingresses {
+		switch ingress.Type {
+		case graph.DirectIngress:
+			err := c.reverseExpandDirect(ctx, &reverseExpandRequest{
+				storeID:          storeID,
+				ingress:          ingress,
+				sourceObjectRef:  sourceObjRef,
+				targetUserRef:    req.User,
+				contextualTuples: req.ContextualTuples,
+				limit:            req.Limit,
+			}, resultChan, foundObjectsMap, foundCount)
+			if err != nil {
+				return err
+			}
+
+		case graph.TupleToUsersetIngress:
+			err := c.reverseExpandTupleToUserset(ctx, &reverseExpandRequest{
+				storeID:          storeID,
+				ingress:          ingress,
+				sourceObjectRef:  sourceObjRef,
+				targetUserRef:    req.User,
+				contextualTuples: req.ContextualTuples,
+				limit:            req.Limit,
+			}, resultChan, foundObjectsMap, foundCount)
+			if err != nil {
+				return err
+			}
+
+		default:
+			return errors.Errorf("unsupported ingress type")
+		}
+	}
+
+	return nil
+}
+
+// StreamedConnectedObjects yields all of the objects of the provided objectType that
+// the given user has a specific relation with. The results will be limited by the request
+// limit. If a 0 limit is provided then all objects of the provided objectType will be
+// returned.
+func (c *ConnectedObjectsCommand) StreamedConnectedObjects(
+	ctx context.Context,
+	req *ConnectedObjectsRequest,
+	resultChan chan<- string, // object string (e.g. document:1)
+) error {
+
+	var foundCount *uint32
+	if req.Limit > 0 {
+		foundCount = new(uint32)
+	}
+
+	var foundObjects sync.Map
+	return c.streamedConnectedObjects(ctx, req, resultChan, &foundObjects, foundCount)
+}
+
+type reverseExpandRequest struct {
+	storeID          string
+	ingress          *graph.RelationshipIngress
+	sourceObjectRef  *openfgapb.RelationReference
+	targetUserRef    *openfgapb.ObjectRelation
+	contextualTuples []*openfgapb.TupleKey
+	limit            uint32
+}
+
+func (c *ConnectedObjectsCommand) reverseExpandTupleToUserset(
```

Comment:
Could we have some tracing/instrumentation in this bit (can be in another PR but please create an issue)

Reply:
Definitely! I was going to submit a follow up PR with more tracing and instrumentation. I want to think through the exact bits I want to capture, but database queries and resolution depths etc.. are probably good starting points.

Reply:
https://github.com/openfga/openfga/issues/280